src/data_source/storage/sql.rs

```diff
@@ -672,38 +782,80 @@ impl VersionedDataSource for SqlStorage {
 // These tests run the `postgres` Docker image, which doesn't work on Windows.
 #[cfg(all(any(test, feature = "testing"), not(target_os = "windows")))]
 pub mod testing {
+    #![allow(unused_imports)]
+    use crate::data_source::storage::sql::sqlx::ConnectOptions;
     use async_compatibility_layer::art::async_timeout;
     use async_std::net::TcpStream;
+    use rand::Rng;
+    use refinery::Migration;
+    use sqlx::sqlite::SqliteConnectOptions;
     use std::{
-        env,
+        env, fs,
+        path::Path,
         process::{Command, Stdio},
-        str,
+        str::{self, FromStr},
         time::Duration,
     };
 
     use portpicker::pick_unused_port;
-    use refinery::Migration;
 
-    use super::Config;
+    use super::{Config, Db};
     use crate::testing::sleep;
 
     #[derive(Debug)]
     pub struct TmpDb {
+        #[cfg(not(feature = "embedded-db"))]
         host: String,
+        #[cfg(not(feature = "embedded-db"))]
         port: u16,
+        #[cfg(not(feature = "embedded-db"))]
         container_id: String,
+        #[cfg(feature = "embedded-db")]
+        db_path: std::path::PathBuf,
+        #[allow(dead_code)]
         persistent: bool,
     }
     impl TmpDb {
+        #[cfg(feature = "embedded-db")]
+        fn init_sqlite_db(persistent: bool) -> Self {
+            let file = tempfile::Builder::new()
+                .prefix("sqlite-")
+                .suffix(".db")
+                .tempfile()
+                .unwrap();
+
+            if persistent {
+                let (_, db_path) = file.keep().unwrap();
+
+                return Self {
+                    db_path,
+                    persistent,
+                };
+            }
+
+            Self {
+                db_path: file.into_temp_path().to_path_buf(),
```

Comment:
I think the `TempPath` ends up getting dropped here, which should delete the file. I thought we would need to store the `TempPath` object in `Self` so that the file doesn't get deleted until `Self` is dropped. 

According to [the docs](https://docs.rs/tempfile/latest/tempfile/struct.TempPath.html) there is a way (`keep(bool)`) to construct a `TempPath` such that it does or doesn't get deleted on drop based on a runtime parameter, maybe we can make use of that somehow